### ISSUE TITLE
Improve reliability of profile website verification and/or implement alternative methods

### ISSUE THREAD

**Issue by fractalcounty** [mastodon/mastodon#33025]
### Pitch

In my opinion, Mastodon's approach towards website ownership verification for additional links on profiles is poorly documented and very unreliable compared to other implementations of site ownership. My main reasoning for this proposal essentially boils down to the fact that there are many very different reasons why site verification can fail, but zero real indication to the end user of if it even failed in the first place. 

While it seems to work for many users, I think the process could be refined in one or more key ways:

- Offering alternative, more reliable methods of site ownership verification (such as Bluesky which [uses a DNS TXT record](https://bsky.social/about/blog/4-28-2023-domain-handle-tutorial), or Discord which allows _both_ [use of a `.well-known/discord` file on your website or a DNS record](https://x.com/DiscordPreviews/status/1679252635700920321?lang=en)). I'd imagine this is the easiest and most effective way to improve the verification process- just offering a plan B for cases where the current tag method just isn't working (or isn't feasible/possible) would already be a massive improvement. 
- Enabling a way to explicitly request verification that provides some form of input in response (e.g status messages), or some indication of what Mastodon is actually seeing and expecting during the verification process
- Providing basic error catching for common types of issues. For example, a simple indication that Mastodon can't access the linked page in the first place would be insanely helpful in discerning whether or not Cloudflare is blocking whatever user agent Mastodon uses as opposed to a simple misconfiguration of the elements in the site's HTML. 
- Generally improving the site verification documentation, including common pitfalls and common areas of user error and misconfiguration

I think some combination of these points would greatly improve the reliability and usability of the site verification process.

### Motivation

My motivation for this feature request is based on my personal experience of verifying my simple, static HTML/CSS website on my Mastodon profile. For illustrative purposes, I'll outline what the entire process looked like for me from start to finish.
 
First, I added a hidden <a> tag containing `rel=me` pointing to my Mastodon profile's URL in my website's body exactly as the verification page described, as well as an additional <link> element in the website's head for redundancy. I deployed the website to Cloudflare Pages via my domain and verified that the expected tags were present and accessible in the public website's raw HTML index page source:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <link rel="me" href="https://mastodon.social/@fractalcounty">
    <!-- ...etc -->
  </head>
  <body>
    <div class="sr-only">
      <a href="https://mastodon.social/@fractalcounty" rel="me">
        Mastodon
      </a>
    </div>
    <!-- ...etc -->
  </body>
</html>
...

I then added the full URL as an additional link on my Mastodon profile (ensuring no typos, i.e trailing slashes) and saved the changes. I waited a few minutes for it to verify, but nothing happened. Since I had 100% verified that the public webpage's source absolutely had the correct elements, I suspected cache was to blame. I enabled Development Mode for my domain in Cloudflare, purged the cache, and periodically tried re-adding the link to my profile over the period of a few days. No luck.

At this point, I felt like I had ruled out simple user error and moved onto more complicated causes. I pawed through my website's source code and ensured that my site's HTML contents weren't being altered or minified in the early stages of the initial request. I tried every variation of ordering the arguments in the <a> and <link> tag's elements but still came up empty handed in terms of verification. Eventually, I had just assumed the verification process took awhile and gave up.

A week or two later, I took another crack at it. This time, I went through every possible avenue of potential misconfiguration on the Cloudflare side of things. I ensured the Cloudflare security level was set to 'Essentially Off', disabled features like Bot Fight Mode, and disabled any features I could. Still no luck. It was at this point that I had the idea to try adding my website's Cloudflare Pages development URL to mastodon (i.e `https://[slug].site-[id].pages.dev/` instead of `https://mydomain.com`). To my surprise, it was verified the second I pressed submit. Assuming this was a smoking gun of some sort of cache related issue (as the cloudflare pages URL has zero caching by default), I took even more measures to rule out Cloudflare as the culprit. I added cache and firewall bypass rules on Cloudflare for potential verification-related hostnames, user agents, et cetera (i.e `mastodon`, `http.rb`, et cetera)- nothing. I setup a subdomain with a CNAME DNS alias pointing towards my Cloudflare Pages site (i.e `verification.mywebsite.com`)- still nothing. 

At this point, I just kept throwing whatever I could at the wall and seeing what stuck, breaking many aspects of my website in the process. A good month after initially trying to get my site verified, I employed extreme cache-busting measures on my website as a last-ditch effort before throwign in the towel. These were (but not limited to) serving a `_headers` file w/ cache control TTL 0, adding an event listener to force a cache refresh for clients, and adding randomized slugs to site's all of my static assets. Then, seemingly out of the blue, it finally worked and my website was verified. 

To me, this process could've been greatly streamlined or simplified if Mastodon simply offered alternative methods of site verification. I think I just ran into a corner case in which the current method of site verification was just flat out incompatible with my website for whatever reason. At the very least, some form of indication of why verification was failing would've been insanely useful. I feel like most people who would run into similar issues would just give up after clicking submit a few times and not seeing the verification checkmark pop up.

**Comment by mjankowski** [comment]
I think it'd be worth attempting the various "more feedback" things first before exploring alternate methods...

- Letting people know we cant access/parse the page
- That we wound the link but not with rel=me
- That we found some rel=me but they were not the link
- How often we try to verify?
- The last verification attempt (and next one?)
- A history of verification attempts and their result?

**Comment by cgay** [comment]
Throwing this in here as it seems related, rather than opening a new bug....

The experience I had while trying to verify the https://opendylan.org website for https://fosstodon.org/@DylanLang was very confusing.  In the end I got help from another user and the problem was that opendylan.org's nginx instance was not using the fullchain.pem certificate file, so curl (and perhaps other tools?) failed.  See https://fosstodon.org/@DylanLang/117107234261499113 for the discussion.

In short, it would have been really helpful if the help text for "How do I add a verified link?" and for "How do I make the link invisible?" explained this full chain requirement or at a minimum said something like "`curl <your URL>` must work".  I'm no cert expert and I assumed that if it works in a browser it should verify.  Turns out browsers are more lenient than curl.

Thanks.
